meson.build

```diff
@@ -129,170 +129,207 @@ thread_dep = dependency('threads') # pthread for tensorflow-lite
+    ],
+    'project_args_when_available': { 'ENABLE_NNFW_RUNTIME': 1 }
+  },
+  'tflite-nnapi-delegation': {
+    'deps': [ dependency('nnfw', required: get_option('tflite-nnapi-delegation')) ],
+    'project_args_when_available': { 'ENABLE_TFLITE_NNAPI_DELEGATION': 1 }
+  },
+  'armnn-support': {
+    'deps': [ dependency('armnn', required: get_option('armnn-support')) ],
+    'project_args_when_available': { 'ENABLE_ARMNN': 1 }
+  },
+  'orcc-support': {
+    'deps': [ orc_dep , pg_orcc ],
+    'project_args_when_available': {'HAVE_ORC': 1},
+    'project_args_when_not_available': { 'DISABLE_ORC': 1 },
+    'extra_args_when_available': {'orcc_args': [pg_orcc, '--include', 'glib.h'] }
+  }
+}
+
+project_args = {}
+# This section controls the flow of feature registration.
+foreach feature_name, data : features
+  _deps = data.get('deps', [])
+  _deps_all_found = true
+
+  foreach dep : _deps
+    if not dep.found()
+      _deps_all_found = false
+    endif
+  endforeach
 
-  if tf_dep.found() and protobuf_dep.found()
-    add_project_arguments('-DENABLE_TENSORFLOW=1', language: ['c', 'cpp'])
-  else
-    error('Cannot find tensorflow')
+  if not _deps_all_found
+    warning('Target dependency is not found for feature @0@'.format(feature_name))
   endif
-endif
 
-# Tensorflow-lite
-if get_option('enable-tensorflow-lite')
-  tflite_dep = dependency('tensorflow-lite', required: true)
-
-  if tflite_dep.found()
-    add_project_arguments('-DENABLE_TENSORFLOW_LITE=1', language: ['c', 'cpp'])
-  else
-    error('Cannot find tensorflow-lite')
+  _is_available = data.get('is_available', _deps_all_found) # falling back to _deps_all_found if is_available is not presen
+  
+  if not _is_available
+    _is_available = data.get('is_available_fallback', false)
   endif
-endif
-
-# PyTorch
-if get_option('enable-pytorch')
-  torch_dep = dependency('pytorch', required: true)
-
-  if torch_dep.found()
-    add_project_arguments('-DENABLE_PYTORCH=1', language: ['c', 'cpp'])
-  else
-    error('Cannot find pytorch')
+  
+  _variable_postfix = _is_available ? '_when_available' : '_when_not_available'
+  
+  # handle project args
+  project_args += data.get('project_args' + _variable_postfix, {})
+
+  # expose variables for further use
+  _vn = feature_name.underscorify()
+
+  set_variable(_vn + '_deps', _deps)
+  set_variable(_vn + '_is_available', _is_available)
+
+  foreach name, value : data.get('extra_args' + _variable_postfix, {})
+    set_variable(_vn + '_' + name, value)
+  endforeach
+
+  # handle_msg and halt with error when not_available and not_enabled at the sametime by definition
+  _feature_status = _is_available ? 'available' : 'unavailable'
+
+  _msg = data.get('message' + _variable_postfix, '@0@ is @1@.'.format(feature_name, _feature_status))
+  if _is_available
+    message(_msg)
+  else # not available
+    get_option(feature_name).enabled() ? error(_msg) : warning(_msg)
```

Comment:
I think you can simplify this code a lot (short lines, simpler logic for the same effects)

Reply:
Revised